src/dragonball/src/vm/mod.rs

```diff
@@ -834,4 +841,204 @@ pub mod tests {
+            },
+            vpmu_feature: 0,
+        };
+
+        let mut vm = create_vm_instance();
+        vm.set_vm_config(vm_config);
+        assert!(vm.init_guest_memory().is_ok());
+        let vm_memory = vm.address_space.vm_memory().unwrap();
+
+        assert_eq!(vm_memory.num_regions(), 1);
+        assert_eq!(vm_memory.last_addr(), GuestAddress(0xffffff));
+
+        // Reconfigure an already configured vm will be ignored and just return OK.
+        let vm_config = VmConfigInfo {
+            vcpu_count: 1,
+            max_vcpu_count: 3,
+            cpu_pm: "off".to_string(),
+            mem_type: "shmem".to_string(),
+            mem_file_path: "".to_string(),
+            mem_size_mib: 16,
+            serial_path: None,
+            cpu_topology: CpuTopology {
+                threads_per_core: 1,
+                cores_per_die: 1,
+                dies_per_socket: 1,
+                sockets: 1,
+            },
+            vpmu_feature: 0,
+        };
+        vm.set_vm_config(vm_config);
+        assert!(vm.init_guest_memory().is_ok());
+        let vm_memory = vm.address_space.vm_memory().unwrap();
+        assert_eq!(vm_memory.num_regions(), 1);
+        assert_eq!(vm_memory.last_addr(), GuestAddress(0xffffff));
+
+        let obj_addr = GuestAddress(0xf0);
+        vm_memory.write_obj(67u8, obj_addr).unwrap();
+        let read_val: u8 = vm_memory.read_obj(obj_addr).unwrap();
+        assert_eq!(read_val, 67u8);
+    }
+
+    #[test]
+    fn test_vm_create_devices() {
+        skip_if_not_root!();
+        let epoll_mgr = EpollManager::default();
+        let vmm = Arc::new(Mutex::new(crate::vmm::tests::create_vmm_instance(
+            epoll_mgr.clone(),
+        )));
+
+        let mut guard = vmm.lock().unwrap();
+        let vm = guard.get_vm_mut().unwrap();
+
+        let vm_config = VmConfigInfo {
+            vcpu_count: 1,
+            max_vcpu_count: 3,
+            cpu_pm: "off".to_string(),
+            mem_type: "shmem".to_string(),
+            mem_file_path: "".to_string(),
+            mem_size_mib: 16,
+            serial_path: None,
+            cpu_topology: CpuTopology {
+                threads_per_core: 1,
+                cores_per_die: 1,
+                dies_per_socket: 1,
+                sockets: 1,
+            },
+            vpmu_feature: 0,
+        };
+
+        vm.set_vm_config(vm_config);
+        assert!(vm.init_guest_memory().is_ok());
+        assert!(vm.setup_interrupt_controller().is_ok());
+
+        let vm_memory = vm.address_space.vm_memory().unwrap();
+        assert_eq!(vm_memory.num_regions(), 1);
+        assert_eq!(vm_memory.last_addr(), GuestAddress(0xffffff));
+
+        let kernel_file = TempFile::new().unwrap();
+        let cmd_line = Cmdline::new(64);
+
+        vm.set_kernel_config(KernelConfigInfo::new(
+            kernel_file.into_file(),
+            None,
+            cmd_line,
+        ));
+
+        vm.init_devices(epoll_mgr).unwrap();
```

Comment:
Should this line use `assert!(xxx.is_ok())` like 970?

Reply:
`xxx.is_ok()` will lose error messages.